components/org.wso2.carbon.crypto.provider.hsm/src/main/java/org/wso2/carbon/crypto/provider/hsm/HSMBasedExternalCryptoProvider.java

```diff
@@ -0,0 +1,675 @@
+
+        String[] keySpecification = symmetricAlgorithm.split("/")[0].split("_");
+        String keyType = keySpecification[0];
+        String errorMessage = String.format("Requested key type generation is not supported for '%s' " +
+                "algorithm", symmetricAlgorithm);
+        SecretKey secretKeyTemplate;
+        switch (keyType) {
+            case (CryptoConstants.KeyType.AES):
+                secretKeyTemplate = KeyTemplateGenerator.generateAESKeyTemplate();
+                ((AESSecretKey) secretKeyTemplate).getValue().setValue(value);
+                break;
+            case (CryptoConstants.KeyType.DES):
+                secretKeyTemplate = KeyTemplateGenerator.generateDESKeyTemplate();
+                ((DESSecretKey) secretKeyTemplate).getValue().setValue(value);
+                break;
+            case (CryptoConstants.KeyType.DES2):
+                secretKeyTemplate = KeyTemplateGenerator.generateDES2KeyTemplate();
+                ((DES2SecretKey) secretKeyTemplate).getValue().setValue(value);
+                break;
+            case (CryptoConstants.KeyType.DES3):
+                secretKeyTemplate = KeyTemplateGenerator.generateDES3KeyTemplate();
+                ((DES3SecretKey) secretKeyTemplate).getValue().setValue(value);
+                break;
+            case (CryptoConstants.KeyType.DESede):
+                secretKeyTemplate = KeyTemplateGenerator.generateDES3KeyTemplate();
+                ((DES3SecretKey) secretKeyTemplate).getValue().setValue(value);
+                break;
+            default:
+                throw new CryptoException(errorMessage);
+        }
+        return generateKey(secretKeyTemplate, false, keyType, session);
+    }
+
+    protected SecretKey generateKey(SecretKey secretKeyTemplate, boolean encryptMode, String keyGenAlgo, Session
+            session) throws CryptoException {
+
+        KeyHandler keyHandler = new KeyHandler(session);
+        if (encryptMode) {
+            return keyHandler.generateSecretKey(secretKeyTemplate, mechanismResolver.resolveMechanism(
+                    new MechanismDataHolder(ENCRYPT_MODE, keyGenAlgo)));
+        } else {
+            return (SecretKey) keyHandler.getKeyHandle(secretKeyTemplate);
+        }
+    }
+
+    protected HybridEncryptionOutput generateHybridEncryptionOutput(Mechanism symmetricMechanism, byte[] encryptedData,
+                                                                    byte[] encryptedKey) throws CryptoException {
+
+        Parameters paramObject = symmetricMechanism.getParameters();
+        if (paramObject instanceof GcmParameters) {
+            CK_GCM_PARAMS gcmParams = (CK_GCM_PARAMS) paramObject.getPKCS11ParamsObject();
+            GCMParameterSpec gcmParameterSpec = new GCMParameterSpec((int) gcmParams.ulTagBits, gcmParams.pIv);
+            int tagPos = encryptedData.length - (int) (gcmParams.ulTagBits) / 8;
```

Comment:
Can you add a comment explaining this logic

Reply:
Fixed in https://github.com/wso2-extensions/carbon-crypto-provider-hsm/pull/5